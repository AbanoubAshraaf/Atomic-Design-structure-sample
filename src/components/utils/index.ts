export const isValidAge = (age:number): boolean => {
    return (
      age > 0 && // Age must be positive
      !isNaN(age) && // Age must be a valid number
      Number.isInteger(age) // Age must be an integer
    );
  };
  