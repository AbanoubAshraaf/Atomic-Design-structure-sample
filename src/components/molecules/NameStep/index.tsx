import React, { useState } from "react";
import { StepProps } from "@/components/types";
import InputField from "@/components/atoms/InputField";
import Button from "@/components/atoms/Button";

const NameStep: React.FC<StepProps> = ({ onClick }) => {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [error, setError] = React.useState<string>("");

  const validateAndProceed = () => {
    if (!firstName || !lastName) {
      setError("Both first and last names are required.");
    } else {
      setError("");
      onClick("name", { firstName, lastName });
    }
  };

  return (
    <div className="flex flex-col gap-4 w-full">
      <InputField
        label="First Name"
        value={firstName}
        type="text"
        onChange={(value) => setFirstName(value)}
      />
      <InputField
        label="Last Name"
        value={lastName}
        type="text"
        error={error}
        onChange={(value) => setLastName(value)}
      />
      <Button text="Next" onClick={validateAndProceed} />
    </div>
  );
};

export default NameStep;
