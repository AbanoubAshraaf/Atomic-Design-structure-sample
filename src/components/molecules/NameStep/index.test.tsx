import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { mockValidDesignerUserData } from "@/components/test/mocks";
import { createSnapShot } from "@/components/test/utils";
import NameStep from ".";

describe("<NameStep/> Component tests", () => {
  test("render first name and last name inputs properly", () => {
    const { container, getByTestId } = render(<NameStep onClick={jest.fn()} />);
    expect(getByTestId("first name")).toBeInTheDocument();
    expect(getByTestId("last name")).toBeInTheDocument();

    createSnapShot(container);
  });

  test("display validation error if fields are empty", () => {
    const { container, getByText } = render(<NameStep onClick={jest.fn()} />);
    fireEvent.click(getByText("Next"));
    expect(getByText("Both first and last names are required.")).toBeTruthy();

    createSnapShot(container);
  });

  test("calls callback with name values on button click", () => {
    const mockCallback = jest.fn();
    const { container, getByTestId, getByText } = render(
      <NameStep onClick={mockCallback} />,
    );

    fireEvent.change(getByTestId("first name"), {
      target: { value: mockValidDesignerUserData.name?.firstName },
    });
    fireEvent.change(getByTestId("last name"), {
      target: { value: mockValidDesignerUserData.name?.lastName },
    });
    fireEvent.click(getByText("Next"));

    expect(mockCallback).toHaveBeenCalledWith("name", {
      firstName: "John",
      lastName: "Doe",
    });

    createSnapShot(container);
  });
});
