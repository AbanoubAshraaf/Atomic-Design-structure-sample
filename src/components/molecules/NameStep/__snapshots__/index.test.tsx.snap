// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NameStep/> Component tests calls callback with name values on button click 1`] = `
<div>
  <div
    class="flex flex-col gap-4 w-full"
  >
    <div
      class="flex flex-col gap-2 items-start"
    >
      <label>
        First Name
        :
      </label>
      <input
        class="input input-bordered input-primary w-full max-w-xs"
        data-testid="first name"
        type="text"
        value="John"
      />
    </div>
    <div
      class="flex flex-col gap-2 items-start"
    >
      <label>
        Last Name
        :
      </label>
      <input
        class="input input-bordered input-primary w-full max-w-xs"
        data-testid="last name"
        type="text"
        value="Doe"
      />
    </div>
    <button
      class="btn btn-primary"
    >
      Next
    </button>
  </div>
</div>
`;

exports[`<NameStep/> Component tests display validation error if fields are empty 1`] = `
<div>
  <div
    class="flex flex-col gap-4 w-full"
  >
    <div
      class="flex flex-col gap-2 items-start"
    >
      <label>
        First Name
        :
      </label>
      <input
        class="input input-bordered input-primary w-full max-w-xs"
        data-testid="first name"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex flex-col gap-2 items-start"
    >
      <label>
        Last Name
        :
      </label>
      <input
        class="input input-bordered input-primary w-full max-w-xs"
        data-testid="last name"
        type="text"
        value=""
      />
      <span
        class="error-text"
      >
        Both first and last names are required.
      </span>
    </div>
    <button
      class="btn btn-primary"
    >
      Next
    </button>
  </div>
</div>
`;

exports[`<NameStep/> Component tests render first name and last name inputs properly 1`] = `
<div>
  <div
    class="flex flex-col gap-4 w-full"
  >
    <div
      class="flex flex-col gap-2 items-start"
    >
      <label>
        First Name
        :
      </label>
      <input
        class="input input-bordered input-primary w-full max-w-xs"
        data-testid="first name"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex flex-col gap-2 items-start"
    >
      <label>
        Last Name
        :
      </label>
      <input
        class="input input-bordered input-primary w-full max-w-xs"
        data-testid="last name"
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary"
    >
      Next
    </button>
  </div>
</div>
`;
